### PR TITLE
Post checkout

### PR DESCRIPTION
**Pull request by pyepye**
I wrote this into my own app originally but thought it was probably helpful to everyone so made it a little more generic and move it over.

* Reusable template to easily add PaddleJS and `Paddle.Setup` to a page
* Checkout model and a post checkout view to save basic Checkout information
* Some JS to automatically send data to the above API on checkout complete
* Doc update with the above info
* Make `on_subscription_created` use `update_or_create`. This is incase someone wants to create a subscription before a Webhook is received

### CODE REVIEW

File: djpaddle/templates/djpaddle_post_checkout.html [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                encodedString += '&';
+            }
+            encodedString += encodeURI(prop + '=' + data[prop]);
+        }
+    }
+    return encodedString;
+}
+
+function addCheckoutCompleteToPaddleButtons() {
+    paddleButtons = document.getElementsByClassName("paddle_button");
```
Comment (FPurchess) [inline djpaddle/templates/djpaddle_post_checkout.html:42]: `var` missing?

File: docs/paddle_checkout.rst [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    def PaddlecheckoutReciever(sender, instance, created, **kwargs):
+        if created:
+            # Get checkout and subscription data here
+
+
+    post_save.connect(PaddlecheckoutReciever, sender=Checkout)
+
+
+Another option is a background task to compare each ``Checkout.id`` against each ``Subscription.checkout_id`` to ensure no Webhooks have been missed.
```
Comment (FPurchess) [inline docs/paddle_checkout.rst:73]: @pyepye do you think it makes sense to create a management-command for that as well?

File: docs/paddle_checkout.rst [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+Other post checkout options
+---------------------------
+
+If you want to manually configure what your happens after a checkout has been completed instead of using the ``checkout_push.html`` template please see:
```
Comment (FPurchess) [inline docs/paddle_checkout.rst:79]: `what your happens` ==> `what happens`

File: tests/test_checkout_api.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+from urllib.parse import urlencode
+
+import pytz
+from django.test import Client, TestCase
+from django.urls import reverse
+
+from djpaddle.models import Checkout
+
+
+class TestWebhook(TestCase):
```
Comment (FPurchess) [inline tests/test_checkout_api.py:11]: Should be `CheckoutTestCase`, no?

**Comment by pyepye** [comment]
@FPurchess  I've fixed all those bits up thanks.

I also did some testing with Paddle's redirects and it didn't quite work as expected so I've added in a way to redirect with `djpaddle_post_checkout.html` and some info in the docs about it to try and save some people the same hassle.

I've also just made a note that a few management commands are coming soon.

File: djpaddle/templates/djpaddle_post_checkout.html [hunk 5]
```diff
@@ -19,27 +14,49 @@
...
+            console.log(xhr.status);
+            if (xhr.status === 200) {
+                var jsonResponse = JSON.parse(xhr.responseText);
+                if (jsonResponse.hasOwnProperty("redirect_url")) {
+                    window.location.href = jsonResponse["redirect_url"];
+                }
+            } else if (xhr.status === 204) {
+                console.log("Checkout complete with ID: " + data.checkout.id)
+            } else {
+                alert("There was a problem with the request.");
```
Comment (FPurchess) [inline djpaddle/templates/djpaddle_post_checkout.html:38]: @pyepye this should rather redirect to an error-page

File: djpaddle/templates/djpaddle_post_checkout.html [hunk 6]
```diff
@@ -19,27 +14,49 @@
...
 
+    var xhr = new XMLHttpRequest();
+    xhr.open("POST", "{% url 'djpaddle:post_checkout_api' %}?next={{ djpaddle_checkout_success_redirect }}");
+    xhr.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
+    xhr.setRequestHeader("X-CSRFToken", csrftoken);
+
+    xhr.onreadystatechange = function(){
+      // Process the server response here.
+        if (xhr.readyState === XMLHttpRequest.DONE) {
+            console.log(xhr.status);
```
Comment (FPurchess) [inline djpaddle/templates/djpaddle_post_checkout.html:29]: @pyepye please remove `console.log` or only show them when `DEBUG=True`

File: docs/paddle_checkout.rst [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+
+2. Using a ``post_save`` receiver on the ``Checkout`` model and using the `Paddle API getorder <https://developer.paddle.com/api-reference/checkout-api/order-information/getorder>`_ endpoint
+
+.. code-block:: python
+
+    from djpaddle.models import Checkout, Subscription
+
+
+    def PaddlecheckoutReciever(sender, instance, created, **kwargs):
```
Comment (FPurchess) [inline docs/paddle_checkout.rst:99]: @pyepye any reason why you did not use snake_case?

File: djpaddle/models.py [hunk 8]
```diff
@@ -201,19 +201,29 @@ def __str__(self):
         return "Subscription <{}:{}>".format(str(self.subscriber), str(self.id))
 
 
+class Checkout(models.Model):
+    # Backup model used to store checkout info from PaddleJS. Acts as
```
Comment (FPurchess) [inline djpaddle/models.py:205]: @pyepye Let's use tripple-quotes here (""") to be consistent on how to comment classes

File: docs/paddle_checkout.rst [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+If you want to redirect the user after the checkout process is complete while using ``djpaddle_post_checkout.html`` you can add a  context variable called ``djpaddle_checkout_success_redirect``.
+
+This will then redirect the user after the checkout data has been saved. The redirect will also include a checkout query parameter ``?checkout={checkout_id}`` so you can look up the checkout information on the redirected page.
+
+
+Post-Checkout Order Information
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once the user has completed an order you should there are several ways you can then get order details.
```
Comment (FPurchess) [inline docs/paddle_checkout.rst:71]: @pyepye is there something missing in this sentence?

File: djpaddle/templates/djpaddle_post_checkout.html (hunk 5 above)
Comment (pyepye) [reply djpaddle/templates/djpaddle_post_checkout.html]: I'm not sure it's a good idea to redirect to any error page as it's not the Paddle checkout that has failed. It's just the dj-paddle post checkout request which isn't _that_ important as a webhook should come along soon and ensure we know the checkout was a success.

**Comment by pyepye** [comment]
Thanks for the review again. I've cleaned up those bits.
Sorry about the poor English in the docs, you would never guess it was my first language.

**Pull request merged**